dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,207 @@
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+    PARAM_CHECKSUM = "md5Checksum"
+    DEFAULT_GOOGLE_AUTH_SETTINGS_PATH = os.path.join(
+        os.path.dirname(__file__), "settings.yaml"
+    )
+    FOLDER_MIME_TYPE = "application/vnd.google-apps.folder"
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        if Config.SECTION_GDRIVE_CREDENTIALPATH not in config:
+            shared_token_warning()
+        self.gdrive_credentials_path = config.get(
+            Config.SECTION_GDRIVE_CREDENTIALPATH,
+            self.DEFAULT_GOOGLE_AUTH_SETTINGS_PATH,
+        )
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.init_drive()
+
+    def init_drive(self):
+        self.get_path_id(self.path_info, create=True)
+
+    def list_drive_item(self, query):
+        for page in self.drive.ListFile({"q": query, "maxResults": 1000}):
+            for item in page:
+                yield item
+
+    @cached_property
+    def cached_root_dirs(self):
+        cached_dirs = {}
+        for dir1 in self.list_drive_item(
+            "'%s' in parents and trashed=false" % self.path_info.netloc
+        ):
+            cached_dirs[dir1["title"]] = dir1["id"]
+        return cached_dirs
+
+    @cached_property
+    def raw_drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+        import logging
+
+        if os.getenv("PYDRIVE_USER_CREDENTIALS_DATA"):
+            with open("credentials.json", "w") as credentials_file:
+                credentials_file.write(
+                    os.getenv("PYDRIVE_USER_CREDENTIALS_DATA")
+                )
+
+        # Supress import error on GoogleAuth warning
+        logging.getLogger("googleapiclient.discovery_cache").setLevel(
+            logging.ERROR
+        )
+
+        GoogleAuth.DEFAULT_SETTINGS["client_config_backend"] = "settings"
+        gauth = GoogleAuth(settings_file=self.gdrive_credentials_path)
+        gauth.CommandLineAuth()
+        return GoogleDrive(gauth)
+
+    @property
+    @ratelimit.sleep_and_retry
+    @ratelimit.limits(calls=10, period=1)
+    def drive(self):
+        return self.raw_drive
+
+    def create_drive_item(self, parent_id, title):
+        item = self.drive.CreateFile(
+            {
+                "title": title,
+                "parents": [{"id": parent_id}],
+                "mimeType": self.FOLDER_MIME_TYPE,
+            }
+        )
+        item.Upload()
+        return item
+
+    def get_drive_item(self, name, parent_id):
+        return next(
+            iter(
+                self.drive.ListFile(
+                    {
+                        "q": "'%s' in parents and trashed=false and title='%s'"
```

Comment:
Let's move this out of the block and use `format` instead of `%`, no need to cramp everything into one statement :slightly_smiling_face: